### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by srgmh**
Added:
- api_users app
- JWTMiddleware
 
Connected drf-spectacular.

### CODE REVIEW

File: tradeapp_django/api_users/serializers.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+User = get_user_model()
+
+
+class UserSerializer(serializers.ModelSerializer):
+    password = serializers.CharField(write_only=True)
+
+    class Meta:
+        model = User
+        fields = ['id', 'email', 'password', 'role']
+        extra_kwargs = {
```
Comment (FridrixJo) [inline tradeapp_django/api_users/serializers.py:13]: add read_only_fileds. make it in tuple example
(           "id",
            "is_blocked",
            "subscriptions",
            "suitcase",
            "created_at",
            "updated_at"
)

File: tradeapp_django/api_users/services/jwt_middleware.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        path = request.path
+        if path not in self.paths:
+            return self.get_response(request)
+
+        response = check_token(request)
+        if response:
+            return response
+
+        response = self.get_response(request)
+        return response
```
Comment (FridrixJo) [inline tradeapp_django/api_users/services/jwt_middleware.py:42]: make authentication classes to each view which has to be protected. and make class for working with tokens (class with staticmethods)

File: tradeapp_django/api_users/services/token_services.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    expiry_time = datetime.utcnow() + timedelta(minutes=minutes_valid)
+    payload = {
+        'user_id': user_id,
+        'exp': expiry_time,
+    }
+    token = jwt.encode(payload, SECRET_KEY, algorithm='HS256')
+    return token
+
+
+def check_token(request):
```
Comment (FridrixJo) [inline tradeapp_django/api_users/services/token_services.py:25]: make type hints to all function (args and return value)

File: tradeapp_django/api_users/views.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class UserViewSet(viewsets.ViewSet):
+
+    serializer_class = UserSerializer
+
+    @action(methods=['post'], detail=False)
+    def register(self, request):
+        serializer = UserSerializer(data=request.data)
+        if serializer.is_valid():
```
Comment (FridrixJo) [inline tradeapp_django/api_users/views.py:24]: serializer.is_valid(raise_exception=True)

File: tradeapp_django/api_users/views.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                user.id, settings.REFRESH_TOKEN_EXPIRATION_MINUTES)
+
+            return Response({
+                'user_id': user.id,
+                'access_token': access_token,
+                'refresh_token': refresh_token})
+
+        raise AuthenticationFailed('Invalid email or password')
+
+    @action(methods=['get'], detail=False)
```
Comment (FridrixJo) [inline tradeapp_django/api_users/views.py:48]: @action(detail=False, url_path="login", methods=["post"])

File: tradeapp_django/api_users/services/jwt_middleware.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        self.get_response = get_response
+        self.paths = self.include_paths
+
+    def __call__(self, request):
+
+        path = request.path
+        if path not in self.paths:
+            return self.get_response(request)
+
+        response = check_token(request)
```
Comment (FridrixJo) [inline tradeapp_django/api_users/services/jwt_middleware.py:37]: shit middleware 
#         if request.path in target_list:
#             request.user = SafeJWTAuthentication().authenticate(request)

File: tradeapp_django/api_users/views.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            user_id, settings.REFRESH_TOKEN_EXPIRATION_MINUTES)
+
+        return Response({
+            'user_id': user_id,
+            'access_token': access_token,
+            'refresh_token': refresh_token
+        })
+
+    @action(methods=['get'], detail=False)
+    def profile(self, request):
```
Comment (FridrixJo) [inline tradeapp_django/api_users/views.py:69]: this work has to be done by middleware. middleware authenticates the user

File: tradeapp_django/api_users/serializers.py (hunk 1 above)
Comment (srgmh) [reply tradeapp_django/api_users/serializers.py:13]: done

File: tradeapp_django/api_users/services/jwt_middleware.py (hunk 6 above)
Comment (srgmh) [reply tradeapp_django/api_users/services/jwt_middleware.py:37]: reworked diz SHIT
I hope u like it more

File: tradeapp_django/api_users/views.py (hunk 4 above)
Comment (srgmh) [reply tradeapp_django/api_users/views.py:24]: nice idea, done

File: tradeapp_django/api_users/views.py (hunk 7 above)
Comment (srgmh) [reply tradeapp_django/api_users/views.py:69]: deleted

File: tradeapp_django/api_users/views.py (hunk 5 above)
Comment (srgmh) [reply tradeapp_django/api_users/views.py:48]: added to actions

File: tradeapp_django/api_users/services/token_services.py (hunk 3 above)
Comment (srgmh) [reply tradeapp_django/api_users/services/token_services.py:25]: added

**Pull request merged**